_posts/drafts/exploring-go-objects-part2.md

```diff
@@ -1,76 +1,411 @@
+        return 0, nil
 }
 
-func useSomeProtocol(s SomeProtocol) {
-        s.Do("hi")
-        s.DoOther(5)
+func (f *File) Close() error {
+        return nil
+}
+
+func useFileProtocol(open Opener) {
+        f, _ := open()
+        data := make([]byte, 50)
+        f.Read(data)
+        f.Close()
 }
 
 func main() {
-        useSomeProtocol(SomeProtocol{
-            Do: func(a string) error {
-                    return nil
-            },
-            DoOther: func(a int) (int, error) {
-                    return 0, nil
-            },
+        useFileProtocol(func() (ReadCloser, error) {
+                return &File{}, nil
         })
 }
 ```
 
-The negative point of this is that there is no easy way to verify at
-compile time that all the functions of the interface instance
-have been initialized with a proper function.
+One key difference is that with interfaces this code is now
+safe. The **useFileProtocol** do not have to worry with calling
+functions that are actually nil, the Go compiler will do the work
+of creating a struct that holds a pointer with a **iface** descriptor
+that has all the functions required to satisfy the protocol.
+It does this per match of type <-> interface, as it is used (initializes
+on first usage). 
+
+You still can cause a segmentation fault if you do something like this:
+
+```go
+func() (ReadCloser, error) {
+        var a ReadCloser
+        return a, nil
+}
+```
+
+But there is still the bonus that every time you initialize the variable
+properly, like this :
+
+```go
+func() (ReadCloser, error) {
+        var a ReadCloser = &Whatever{}
+        return a, nil
+}
+```
+
+If it compiles you can be sure it is safe to call all functions of the interface.
+Also with the Go's interface mechanism an object can implement multiple different
+protocols without even knowing these protocols.
+
+How useful can be implementing protocols that you do not even known exists ?
+This is very useful if you want your code to be truly extensible. Allow me
+to provide a real world example from [nash](https://github.com/NeowayLabs/nash).
+
+# Extending code beyond its original purpose
+
+The first time I understood how powerful Go interfaces can be was
+when I was trying to code the tests for the built-in function **exit**
+on nash. The main problem was that it seemed like we would have to
+implement different tests for each platform because the exit status
+code was handled differently on some platforms. I don't remember all
+the details right now but on plan9 the exit status is an string instead
+of an integer.
+
+Basically on an error I wanted the status code, not just the error like
+it is provided on [Cmd.run](https://golang.org/pkg/os/exec/#Cmd.Run).
+
+There is the [ExitError](https://golang.org/pkg/os/exec/#ExitError) type,
+so I could do something like this:
+
+```go
+if exiterr, ok := err.(*exec.ExitError); ok {
+}
+```
+
+To at least known that this is some error generated by an error
```

Comment:
I think here is 'know' instead of 'known'.

Reply:
thanks :D